Quick mode

harendra-kumar opened the pull request:
This is a work in progress based on the yet unmerged getrusage branch. I will rebase it when that gets merged.

harendra-kumar committed:
Make runBenchMark more flexible

runBenchmark now accepts the threshold of sample duration and how many samples
we must take which are above the threshold. This will allow us to use
runBenchmark in more flexible ways.

harendra-kumar committed:
refactor runMode

vincenthz commented on Gauge/Analysis.hs:
```diff
@@ -141,6 +144,53 @@ scale f s@SampleAnalysis{..} = s {
+-- us to have some trust in the raw measurement. We must take minimum 10
+-- samples above this threshold to perform a meaningful statistical analysis.
+threshold :: Int
+threshold = 30
```
Untyped Int and Double ought to not be used anywhere. This really should be a better type to prevent adding different units together.

harendra-kumar replied:
Yes, In fact I almost made a mistake when using this.

harendra-kumar replied:
I am not changing it in this changeset. We should change the ones in `Measured`, this one and maybe other time quantities at the same time. All time quantities should use consistent units.

harendra-kumar committed:
Fix merge issue

harendra-kumar committed:
Implement quick mode

In this mode:

1) Speed is roughly 10x faster e.g. 5 secs vs 50 secs. No more waiting for
   results.
2) There is no statistical analysis of samples done
3) We take 2 samples above 30ms duration and print the stats for the last one.
   We do not take avg, min or max. We print just one sample because all the
   stats can be consistently interpreted with respect to each other. That is
   not possible when we take averages.
4) In default mode only the time is printed, in verbose mode all stats are
   printed. In future we can have a command line selection filter to select
   which stats are to be printed.
5) Results are quite stable, there is a little more variance compared to
   statistically analysed results but it is not significant. The bang for the
   buck is quite high.

harendra-kumar committed:
Add tests for quick and quick verbose modes

harendra-kumar committed:
Access vector safely when it is empty

harendra-kumar committed:
Do not use stime and utime when they are not filled

harendra-kumar commented:
@vincenthz this is ready for review and merge. Commit https://github.com/vincenthz/hs-gauge/pull/13/commits/db265d027814f1e879aeb02d89874c868204bb8c has a summary of the overall change.

harendra-kumar commented:
There seems to be some travis issue for os x tests. Anyway I have tested it on my mac, so it should be okay.
